variod.c

```diff
@@ -339,22 +340,55 @@ int main(int argc, char *argv[])
 
 				} else {
 					fprintf(stderr, "send failed\n");
+                                        // we can skip a pendig request since XCSoar is not online
+                                        request_current_settings = false;
 				}
 			}	
 			// check if there is communication from XCSoar to us
-			if ((read_size = recv(xcsoar_sock , client_message , 2000, 0 )) > 0 )
+			while ((read_size = recv(xcsoar_sock , client_message , 2000, 0 )) > 0 )
 			{
 				// we got some message
 				// terminate received buffer
 				client_message[read_size] = '\0';
 				
 				ddebug_print("Message from XCSoar: %s\n",client_message);
 
+                                char *p = client_message;
+                                // chop one message into multiple lines
+                                while (*p) {
+                                  if (*p == '\n' || *p == '\r') {
+                                    *p = '\0';
+                                    }
+                                  p += 1;
+                                  }
+
+                                int c_cntr = read_size;
+                                char *next_sentence = client_message;
+                                p  = client_message;
 				// parse message from XCSoar
-				parse_NMEA_command(client_message);	
-        	
+                                // one sentence at a time
+                                do {
+                                    parse_NMEA_command(next_sentence);	
+                                    p += (strlen(next_sentence) + 1);
+                                    c_cntr -= (strlen(next_sentence) + 1);
+                                    while ((c_cntr > 1) && (*p == '\0')) {
+                                      c_cntr -= 1;
+                                      p += 1;
+                                      }
+                                    if (c_cntr > 1) next_sentence = p;
+                                    else next_sentence = NULL;
+                                  } while (next_sentence);
+
 			}
-			
+
+                        // we might see if we need to update settings from XCSoar
+                        if (request_current_settings) {
+                          strcpy(client_message,"$POV,?,RPO,MC*"); // all we need for STF
+                          add_checksum(client_message);
+                          send(xcsoar_sock, client_message, strlen(client_message), 0);
+                          // we have triggered to send the data, now we can cancel the reques
+                          request_current_settings = false;
```

Comment:
What if we don't receive the response? Are we going to request it again?

Reply:
the flag is set before the connection can be established. And it doesn't establish unless XCSoar and sensord are connected. I've verified this on the test bench,
I will make the error messages more descriptive. `failed to connect, trying again` etc will tell more what it's doing. That should help. Tomorrow.
We need someone to take a really close look at sensord.